rb/spec/integration/selenium/webdriver/bidi_spec.rb

```diff
@@ -52,6 +52,32 @@ module WebDriver
           level: BiDi::LogInspector::LOG_LEVEL[:ERROR]
         )
       end
+
+      it 'does not close BiDi session if one window is closed' do
+        status = driver.bidi.session.status
+        expect(status.ready).to be false
+        expect(status.message).to be_a String
+
+        driver.switch_to.new_window(:window)
+        driver.switch_to.new_window(:tab)
+        driver.switch_to.new_window(:tab)
+
+        driver.close
+
+        status_after_closing = driver.bidi.session.status
+        expect(status_after_closing.ready).to be false
+        expect(status_after_closing.message).to be_a String
+      end
+
+      it 'closes BiDi session if last window is closed' do
```

Comment:
@titusfortner this tests works in chrome but gives error in firefox. And the error is not in the `it` block but in the `after` block where it calls quit_driver.

This is the error:
```
Selenium::WebDriver::Error::InvalidSessionIdError:
       Tried to run command without establishing a connection 
```
This is not happening in chrome. The after block works fine in it. But in firefox it gives the error. Can you suggest why?

Reply:
I would assume that Firefox closes the connection and the follow-up attempts to quit driver raise exceptions. Chrome might be more at ease and not raise anything at all.

It would help if you run the spec with `DEBUG=1` and paste the output.

Reply:
```
2023-01-21 14:30:17 DEBUG Selenium WebSocket -> {:method=>"session.status", :params=>{}, :id=>1}
2023-01-21 14:30:17 DEBUG Selenium WebSocket <- {"id"=>1, "result"=>{"ready"=>false, "message"=>"Session already started"}}
2023-01-21 14:30:17 INFO Selenium -> DELETE session/77e7158e-b1df-44f5-9f18-8ba0f7b55e3c/window
1674311417155	Marionette	INFO	Stopped listening on port 61474
console.error: Region.jsm: "Error fetching region" (new TypeError("NetworkError when attempting to fetch resource.", ""))
console.error: Region.jsm: "Failed to fetch region" (new Error("NO_RESULT", "resource://gre/modules/Region.jsm", 419))
2023-01-21 14:30:17 DEBUG Selenium    <<<  {"content-type"=>["application/json; charset=utf-8"], "cache-control"=>["no-cache"], "content-length"=>["12"], "date"=>["Sat, 21 Jan 2023 14:30:17 GMT"]}
2023-01-21 14:30:17 INFO Selenium <- {"value":[]}
2023-01-21 14:30:17 DEBUG Selenium WebSocket -> {:method=>"session.status", :params=>{}, :id=>2}
2023-01-21 14:30:17 INFO Selenium -> DELETE session/77e7158e-b1df-44f5-9f18-8ba0f7b55e3c
2023-01-21 14:30:17 DEBUG Selenium    <<<  {"content-type"=>["application/json; charset=utf-8"], "cache-control"=>["no-cache"], "content-length"=>["123"], "date"=>["Sat, 21 Jan 2023 14:30:17 GMT"]}
2023-01-21 14:30:17 INFO Selenium <- {"value":{"error":"invalid session id","message":"Tried to run command without establishing a connection","stacktrace":""}}
```

Here's what I think happens. When I call `driver.close` (line 77) with only one window open, firefox closes the window and calls quit to close the session. That's where we get `Stopped listening on port 61474` in the above log.

But after this, firefox again tries to `quit_driver` (irrespective if I put after block or not, maybe it does in the spec_helper) and that gives the last error in the log.

Chrome might be easy on this check as you said. In the sense that if it calls `quit_driver` when the session is already closed, it doesn't throw error.

A solution that I found is adding `create_driver!` after my expect statement (line 79). This creates a new driver and thus `quit_driver` works, both in firefox and chrome. 

Shall I add that?? @p0deje

Reply:
I think the simplest thing you could do is to rescue `Selenium::WebDriver::Error::InvalidSessionIdError` and ignore it in `after` hook.